File: Assignment 2/climate.java
import java.util.*;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.nio.file.*;
import java.io.*;

class ClimateRecord {
    LocalDate date;
    int maxGust;
    float totalPrecipitation;
    float minTemperature;
    float maxTemperature;
    float avgTemperature;

    public ClimateRecord(LocalDate date, int maxGust, float totalPrecipitation, float minTemperature, float maxTemperature, float avgTemperature) {
        this.date = date;
        this.maxGust = maxGust;
        this.totalPrecipitation = totalPrecipitation;
        this.minTemperature = minTemperature;
        this.maxTemperature = maxTemperature;
        this.avgTemperature = avgTemperature;
    }
}

public class climate {
    public static void main(String[] args) {
        ArrayList<ClimateRecord> records = new ArrayList<>();

        loadRecords(records);
        analyzeData(records);

        while (true) {
            System.out.println("1. Monthly report");
            System.out.println("2. Weather records between two dates");
            System.out.println("3. Exit");
            System.out.println("Enter your choice: ");
            Scanner scanner = new Scanner(System.in);
            String choice = scanner.nextLine();
            scanner.close();

            if (choice.equals("1")) {
                long startTime = System.nanoTime();
                monthlyReport(records);
                long endTime = System.nanoTime();
                long duration = (endTime - startTime);  
                double durationInSeconds = (double)duration / 1_000_000_000;
                System.out.println("Monthly report took " + durationInSeconds + " seconds");
            } else if (choice.equals("2")) {
                System.out.println("Enter the start date (YYYY-MM-DD): ");
                String startDate = scanner.nextLine();
                System.out.println("Enter the end date (YYYY-MM-DD): ");
                String endDate = scanner.nextLine();
                dateRangeReport(records, LocalDate.parse(startDate), LocalDate.parse(endDate));
            } else if (choice.equals("3")) {
                break;
            } else {
                System.out.println("Invalid choice. Please try again.");
            }
        }
    }

    public static void loadRecords(ArrayList<ClimateRecord> records) {
        Path pathToFile = Paths.get("climate-daily.csv");

        try (BufferedReader br = Files.newBufferedReader(pathToFile)) {
            String line = br.readLine();
            while ((line = br.readLine()) != null) {
                String[] attributes = line.split(",");
                if (attributes.length < 6) {
                    System.out.println("Skipping line with missing data: " + line);
                    continue;
                } 
                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd H:mm");
                LocalDate date = LocalDate.parse(attributes[0], formatter);
                try {
                    int maxGust = Integer.parseInt(attributes[1]);
                    float totalPrecipitation = Float.parseFloat(attributes[2]);
                    float minTemperature = Float.parseFloat(attributes[3]);
                    float maxTemperature = Float.parseFloat(attributes[4]);
                    float avgTemperature = Float.parseFloat(attributes[5]);
                    ClimateRecord record = new ClimateRecord(date, maxGust, totalPrecipitation, minTemperature, maxTemperature, avgTemperature);
                    records.add(record);
                } catch (NumberFormatException nfe) {
                    continue;
                }
                
            }
            } catch (IOException ioe) {
                ioe.printStackTrace();
        }
    }

    public static void analyzeData(ArrayList<ClimateRecord> records) {
        Map<String, Float> precipitationPerMonth = new HashMap<>();
        LocalDate maxGustDay = null;
        int maxGust = 0;
        LocalDate maxTempFluctuationDay = null;
        float maxTempFluctuation = 0;
    
        for (ClimateRecord record : records) {
            String monthKey = record.date.format(DateTimeFormatter.ofPattern("MM-yyyy"));
            precipitationPerMonth.put(monthKey, precipitationPerMonth.getOrDefault(monthKey, 0f) + record.totalPrecipitation);
    
            if (record.maxGust > maxGust) {
                maxGust = record.maxGust;
                maxGustDay = record.date;
            }
    
            float tempFluctuation = record.maxTemperature - record.minTemperature;
            if (tempFluctuation > maxTempFluctuation) {
                maxTempFluctuation = tempFluctuation;
                maxTempFluctuationDay = record.date;
            }
        }
    
        Map.Entry<String, Float> maxEntry = null;
        for (Map.Entry<String, Float> entry : precipitationPerMonth.entrySet()) {
            if (maxEntry == null || entry.getValue().compareTo(maxEntry.getValue()) > 0) {
                maxEntry = entry;
            }
        }
    
        System.out.println("\n--Climate Data Analysis--");
        System.out.println("Month with the most precipitation: " + maxEntry.getKey() + " with " + String.format("%.2f", maxEntry.getValue()) + "mm");
        System.out.println("Day with the highest gust: " + maxGustDay + " with " + maxGust + "km/h");
        System.out.println("Day with the highest temperature fluctuation: " + maxTempFluctuationDay + " with " + String.format("%.2f", maxTempFluctuation) + "°C\n");
    }


    public static void monthlyReport(ArrayList<ClimateRecord> records) {
        Map<String, ClimateRecord> monthlyRecords = new HashMap<>();
        for (ClimateRecord record : records) {
            String monthYear = record.date.getMonth().toString() + "-" + record.date.getYear();
            ClimateRecord monthlyRecord = monthlyRecords.getOrDefault(monthYear, new ClimateRecord(record.date, 0, 0, Float.MAX_VALUE, Float.MIN_VALUE, 0));
            if (monthlyRecord.maxGust < record.maxGust) {
                monthlyRecord.maxGust = record.maxGust;
            }
            monthlyRecord.totalPrecipitation += record.totalPrecipitation;
            monthlyRecord.minTemperature = Math.min(monthlyRecord.minTemperature, record.minTemperature);
            monthlyRecord.maxTemperature = Math.max(monthlyRecord.maxTemperature, record.maxTemperature);
            monthlyRecord.avgTemperature += record.avgTemperature;
            monthlyRecords.put(monthYear, monthlyRecord);
        }
    
        for (Map.Entry<String, ClimateRecord> entry : monthlyRecords.entrySet()) {
            ClimateRecord record = entry.getValue();
            System.out.println("Month-Year: " + entry.getKey());
            System.out.println("Max Gust: " + record.maxGust);
            System.out.println("Total Precipitation: " + record.totalPrecipitation);
            System.out.println("Min Temperature: " + record.minTemperature);
            System.out.println("Max Temperature: " + record.maxTemperature);
            System.out.println("Avg Temperature: " + record.avgTemperature / records.size());
        }
    }

    public static void dateRangeReport(ArrayList<ClimateRecord> records, LocalDate startDate, LocalDate endDate) {
        for (ClimateRecord record : records) {
            if (record.date.isAfter(startDate) && record.date.isBefore(endDate)) {
                System.out.println("Date: " + record.date);
                System.out.println("Max Gust: " + record.maxGust);
                System.out.println("Total Precipitation: " + record.totalPrecipitation);
                System.out.println("Min Temperature: " + record.minTemperature);
                System.out.println("Max Temperature: " + record.maxTemperature);
                System.out.println("Avg Temperature: " + record.avgTemperature);
            }
        }
    }
}

File: Assignment 2/climate.py
from collections import defaultdict
import time
class ClimateData:

    def __init__(self, date: str, max_gust: int, total_precipitation: float, min_temperature: float, max_temperature: float, avg_temperature: float):
        self.date = date
        self.max_gust = max_gust
        self.total_precipitation = total_precipitation
        self.min_temperature = min_temperature
        self.max_temperature = max_temperature
        self.avg_temperature = avg_temperature

# Function to parse the climate data from the file
def parseClimateData(filename: str) -> list[ClimateData]:
    data = []

    with open(filename, 'r') as file:
        for line in file:
            try:
                date, max_gust, total_precipitation, min_temperature, max_temperature, avg_temperature = line.split(',')

                data.append(ClimateData(date, int(max_gust), float(total_precipitation), float(min_temperature), float(max_temperature), float(avg_temperature)))
            except ValueError:
                # dont include lines that are not formatted correctly
                continue
    
    return data

# Function to convert the date from the format YYYY-MM-DD to a more readable format
def convertDateTime(date: str, type: str) -> str:
    months = {"01": "January", "02": "February", "03": "March", "04": "April", "05": "May", "06": "June", "07": "July", "08": "August", "09": "September", "10": "October", "11": "November", "12": "December"}
    if type == "day":
        return f"{months[date.split('-')[1]]} {date.split('-')[2].split(' ')[0]} {date.split("-")[0]}"
    elif type == "month":
        return f"{months[date.split("-")[1]]} {date.split("-")[0]}"

# Function to analyze the climate data
def analyzeData(data: list[ClimateData]) -> None:

    max_precipitation = 0
    max_gust_day = None
    max_gust = 0
    max_temp_fluctuation_day = None
    max_temp_fluctuation = 0
    precipitation_per_month = defaultdict(float)

    for value in data:
        month_key = (convertDateTime(value.date, "month"))
        precipitation_per_month[month_key] += value.total_precipitation

        if value.max_gust > max_gust:
            max_gust = value.max_gust
            max_gust_day = value.date

        if value.max_temperature - value.min_temperature > max_temp_fluctuation:
            max_temp_fluctuation = value.max_temperature - value.min_temperature
            max_temp_fluctuation_day = value.date
    
    max_precipitation_month, max_precipitation = max(precipitation_per_month.items(), key=lambda item: item[1])
    print("\n--Climate Data Analysis--")
    print(f"Month with the most precipitation: {max_precipitation_month} with {max_precipitation:.2f}mm")
    print(f"Day with the highest gust: {convertDateTime(max_gust_day, "day")} with {max_gust}km/h")
    print(f"Day with the highest temperature fluctuation: {convertDateTime(max_temp_fluctuation_day, "day")} with {max_temp_fluctuation:.2f}°C\n")

# Function to generate a user generated report
def userGeneratedReport(data: list[ClimateData]) -> None:
    while True:
        print("1. Monthly report")
        print("2. Weather records between two dates")
        print("3. Exit")
        choice = input("Enter your choice: ")

        if choice == '1':
            start_time = time.time()
            monthly_report(data)
            end_time = time.time()
            print(f"Monthly report took {end_time - start_time} seconds to run.")
        elif choice == '2':
            start_date = input("Enter the start date (YYYY-MM-DD): ")
            end_date = input("Enter the end date (YYYY-MM-DD): ")
            date_range_report(data, start_date, end_date)
        elif choice == '3':
            break
        else:
            print("Invalid choice. Please try again.")

# Function to generate a monthly report
def monthly_report(data: list[ClimateData]) -> None:
    # create a dictionary with default values for each month
    months = defaultdict(lambda: {"count": 0, "max_gust": 0, "total_precipitation": 0, "min_temperature": float('inf'), "max_temperature": float('-inf'), "avg_temperature": 0})

    for value in data:
        month = convertDateTime(value.date, "month")
        months[month]["count"] += 1
        months[month]["max_gust"] += value.max_gust
        months[month]["total_precipitation"] += value.total_precipitation
        months[month]["min_temperature"] = min(months[month]["min_temperature"], value.min_temperature)
        months[month]["max_temperature"] = max(months[month]["max_temperature"], value.max_temperature)
        months[month]["avg_temperature"] += value.avg_temperature

    for month, values in months.items():
        print(f"Month: {month}")
        print(f"Average Max Gust: {values['max_gust'] / values['count']:.2f}")
        print(f"Total Precipitation: {values['total_precipitation']:.2f}")
        print(f"Min Temperature: {values['min_temperature']:.2f}")
        print(f"Max Temperature: {values['max_temperature']:.2f}")
        print(f"Mean Temperature: {values['avg_temperature'] / values['count']:.2f}")
        print()

# Function to generate a report for a date range
def date_range_report(data: list[ClimateData], start_date: str, end_date:str):

    for value in data:
        if value.date >= start_date and value.date <= end_date:
            print(f"\nDate: {convertDateTime(value.date, 'day')}")
            print(f"Max Gust: {value.max_gust}km/h")
            print(f"Total Precipitation: {value.total_precipitation}mm")
            print(f"Min Temperature: {value.min_temperature}°C")
            print(f"Max Temperature: {value.max_temperature}°C")
            print(f"Avg Temperature: {value.avg_temperature}°C\n")

if __name__ == "__main__":
    data = parseClimateData("Assignment 2/climate-daily.csv")
    analyzeData(data)
    userGeneratedReport(data)